### DISCUSSION TITLE
Astonished at the poor developer experience.

### DISCUSSION THREAD

**Discussion by DeVmend** [directus/directus#6871] [Archive]
Directus advertises with Typescript. In the official docs, however, every example of the extension (modules, endpoints, etc.) is given with plain Javascript.

Furthermore, there is no support in the form of a CLI to help you create your own endpoints, nor is there a file watcher that restarts the application when changes are made. 

Don't get me wrong, Directus seems to me to be the best headless CMS with tons of configuration and setting options. But the developer experience should not be ignored here.

Should Directus evolve in extensibility even a little bit in the direction of a NestJS developer experience, this could be a real game changer in the CMS market.

**Comment by DeVmend** [comment]
Have I missed something ?

**Reply by benhaynes** [reply]
Hi @Pii2 — while I always appreciate criticism, this post comes off quite thankless.

We have **one** developer on our team at the moment and are working on an open-source budget. You're saying "_Directus seems to me to be the best headless CMS_"... yet comparing us to companies with hundreds of employees and $1B+ valuations.

Open-source software is built in part by the community, so it is important to be as constructive as possible when posting discussions like this. This single post should have been split into several different **actionable** discussions, with specific tasks and recommendations for how we might solve them. As it stands, this simply feels like you're aimlessly venting in a place reserved for meaningful product discussion.

Having spent 17 years working on this platform (and giving it away for free) alongside _passionate_ contributors donating their time, your title alone makes me a bit upset. Just my two cents.

**Comment by kees-tce** [comment]
I think you missed something indeed.

Writing your own endpoints is very well documented, you shouldn't need a CLI for that. Setting up a development environment with nodemon to watch for file changes in your extensions is literally one Google search away.

Not sure what you're looking for, but if you're using something for free, even though it's as well built/extensive as directus is, you shouldn't use terms as "astonished".

If you think the CLI needs additional functionality, or examples could be provided in TS, feel free to contribute.

By the way, asking questions in this discussion area always leads to great answers by the community in my experience.

**Comment by miccoh1994** [comment]
Its actually not that difficult to setup a directus project and then configure rollup to build any extension and reload the instance. Would anyone be interested in putting together a repo of examples, I can contribute to modules and endpoints because I have build quite a few? @benhaynes Would you be open to something like that? I also built my own cli to generate boilerplate for new modules and endpoints.

**Reply by benhaynes** [reply]
> Its actually not that difficult to setup a directus project and then configure rollup to build any extension and reload the instance. Would anyone be interested in putting together a repo of examples, I can contribute to modules and endpoints because I have build quite a few? @benhaynes Would you be open to something like that? I also built my own cli to generate boilerplate for new modules and endpoints.

Absolutely! Though I defer to @rijkvanzanten on actual implementation. :)

**Comment by supermx1** [comment]
Directus is King!!!!!. Thanks @benhaynes you guys literally changed my life and I say it with every ounce of strength in me. @rijkvanzanten is like a mentor. I am a web designer and discovering a free project like Directus has opened my small mind to soo much more. I could even call myself an upcoming developer now...

**Comment by nickrum** [comment]
By the way, we already have a CLI to scaffold all types of extensions. Typescript support is coming very soon.

**Reply by miccoh1994** [reply]
@nickrum oh my, thank you! I never realised you had the extensions-sdk package. Having been through a lot of the codebase I never actually looked under packages xD. I do notice that its existence is not well documented. Thank you

**Reply by nickrum** [reply]
The package is only a few weeks old. That is also why it is still poorly documented.

**Comment by TheAzack9** [comment]
While it might be a bit off topic to this specific thread, i hope that is ok 😅,  do i want to praise the Directus team for how amazingly helpful they have been from the beginning. As a lone developer being instructed to develop a website with an easy-to-use CMS was i and my CEO delighted the second we found Directus and at that point hadn't it reached the first release candidate yet. While looking through different CMS-es has Directus, by a huge margin, been the best one i've ever worked with as a developer. Every part from setting up the models in the cms to fetching data and displaying it on a webpage through GraphQL has been a blast. The times i have reaced a hickup i've been able to contribute with what knowledge i have to better the platform and the times i haven't been able to help much have the developers (@rijkvanzanten especially) been so amazingly helpful and quick to respond to issues. I just wanna say to all of you that you are doing an amazing job and i look forward to seing where the project will head in a few years time. (When it probably becomes one of to go-to systems ;) )

As a free open-source CMS is this definitely something i will recommend every time to any developer in need of a CMS! It has made my life so much easier as a developer having to develop a website!
